File: frontend/app/composables/api/authentication/authentication.ts
import { useApi, setAccessToken } from "../useApi"
import type {
  ForgotPasswordPayload,
  ForgotPasswordResponse,
  RegisterUserPayload,
  RegisterUserResponse, ResetPasswordPayload
} from "@/composables/dto/auth";
import type {ApiResult} from "@/composables/helpers/types/api";
import type { RawAxiosResponseHeaders, AxiosHeaders } from "axios";

export const useAuthentication = () => {
  const api = useApi()

  const CheckEmailExists = async (email: string): Promise<{ exists: boolean; errors?: string[] }> => {
    try {
      const response = await api.post("/base/authentication/email/exist", { email })
      const exists = response.data?.responseMessage === "Email already exists"
      return { exists }
    } catch (error: unknown) {
      const body = error.response?.data
      const errors: string[] = body?.errors?.map((e: { message: string }) => e.message) ?? [
        body?.response_message ?? "Could not verify email"
      ]
      return { exists: false, errors }
    }
  }

  const RegisterUser = async (
    data: RegisterUserPayload
  ): Promise<ApiResult<RegisterUserResponse, RawAxiosResponseHeaders | (RawAxiosResponseHeaders & AxiosHeaders)>> => {
    try {
      const response = await api.post<RegisterUserResponse>("/base/authentication/register", {
        email: data.email,
        password: data.password,
      })

      return { success: true, data: response.data, headers: response.headers }
    } catch (error: unknown) {
      return { success: false, error: error.response?.data?.message || "Registration failed" }
    }
  }

  const LoginUser = async (
    data: RegisterUserPayload
  ): Promise<ApiResult<RegisterUserResponse, RawAxiosResponseHeaders | (RawAxiosResponseHeaders & AxiosHeaders)>> => {
    try {
      const response = await api.post("/base/authentication/login", {
        email: data.email,
        password: data.password,
      })

      return { success: true, data: response.data, headers: response.headers}
    } catch (error: unknown) {
      return { success: false, error: error.response?.data?.response_message || "Login failed" }
    }
  }

  const Logout = async (): Promise<ApiResult<unknown, RawAxiosResponseHeaders | (RawAxiosResponseHeaders & AxiosHeaders)>> => {
    try {
      // No body — the HttpOnly cookie is sent automatically and cleared by the server
      const response = await api.post("/base/authentication/logout")
      return { success: true, data: response.data, headers: response.headers }
    } catch (error: unknown) {
      return { success: false, error: error.response?.data?.response_message || "Logout failed" }
    } finally {
      setAccessToken(null)
      localStorage.removeItem("session")
    }
  }

  const RefreshToken = async (): Promise<boolean> => {
    // No body — the HttpOnly cookie is sent automatically
    if (!localStorage.getItem('session')) return false

    try {
      const { data } = await api.post("/base/authentication/refresh")
      setAccessToken(data.data.access_token)
      return true
    } catch {
      localStorage.removeItem("session")
      return false
    }
  }

  const ForgotPassword = async (
    data: ForgotPasswordPayload
  ): Promise<ApiResult<ForgotPasswordResponse, RawAxiosResponseHeaders | (RawAxiosResponseHeaders & AxiosHeaders)>> => {
    try {
      const response = await api.post("/base/authentication/forgot-password", {
        email: data.email,
      })

      return { success: true, data: response.data, headers: response.headers}
    } catch (error: unknown) {
      return { success: false, error: error.response?.data?.response_message || "Forgot password failed" }
    }
  }

  const ResetPassword = async (
    data: ResetPasswordPayload
  ): Promise<ApiResult<ForgotPasswordResponse, RawAxiosResponseHeaders | (RawAxiosResponseHeaders & AxiosHeaders)>> => {
    try {
      const response = await api.post("/base/authentication/reset-password", {
        password: data.password,
        confirm_password: data.confirm_password,
        token: data.token,
      })
      return { success: true, data: response.data, headers: response.headers}
    } catch (error: unknown) {
      return { success: false, error: error.response?.data?.response_message || "Reset password failed" }
    }
  }

  const RequestEmailCode = async (email: string): Promise<ApiResult<unknown, RawAxiosResponseHeaders | (RawAxiosResponseHeaders & AxiosHeaders)>> => {
    try {
      const response = await api.post("/base/authentication/email-code/request", { email })
      return { success: true, data: response.data, headers: response.headers }
    } catch (error: unknown) {
      return { success: false, error: error.response?.data?.response_message || "Failed to send code" }
    }
  }

  const VerifyEmailCode = async (
    email: string,
    code: string
  ): Promise<ApiResult<RegisterUserResponse, RawAxiosResponseHeaders | (RawAxiosResponseHeaders & AxiosHeaders)>> => {
    try {
      const response = await api.post("/base/authentication/email-code/verify", { email, code })
      return { success: true, data: response.data, headers: response.headers }
    } catch (error: unknown) {
      return { success: false, error: error.response?.data?.response_message || "Invalid or expired code" }
    }
  }

  return {
    CheckEmailExists,
    RegisterUser,
    LoginUser,
    Logout,
    RefreshToken,
    ForgotPassword,
    ResetPassword,
    RequestEmailCode,
    VerifyEmailCode,
  }
}

File: services/base/internal/apps/base/controllers/authentication/authentication.go
package authentication

import (
	"github.com/gofiber/fiber/v2"
	"github.com/sdq-codes/usegro-api/internal/apps/base/dto"
	authenticationService "github.com/sdq-codes/usegro-api/internal/apps/base/services/authentication"
	"github.com/sdq-codes/usegro-api/internal/apps/base/validation"
	"github.com/sdq-codes/usegro-api/internal/helper/auth"
	"github.com/sdq-codes/usegro-api/internal/interface/response"
	"github.com/sdq-codes/usegro-api/pkg/exception"
)

type Controller struct {
	authenticationService authenticationService.Service
}

func NewAuthenticationController(authenticationService authenticationService.Service) *Controller {
	return &Controller{authenticationService: authenticationService}
}

func (uc *Controller) Exist(c *fiber.Ctx) error {
	var req dto.UserExistDTI

	if err := c.BodyParser(&req); err != nil {
		return exception.InvalidRequestBodyError
	}

	authenticationValidation := validation.AuthenticationValidation{}
	if err := authenticationValidation.UserExistValidation(req); err != nil {
		return err
	}

	userExist, err := uc.authenticationService.IsUserEmailExist(c.Context(), req)
	if err != nil {
		return err
	}

	msg := "Email is available"
	if userExist {
		msg = "Email already exists"
	}

	return c.Status(fiber.StatusOK).JSON(response.CommonResponse{
		ResponseCode:    response.RESOURCE_FETCHED,
		ResponseMessage: msg,
	})
}

func (uc *Controller) Register(c *fiber.Ctx) error {
	var req dto.RegisterUserDTI
	if err := c.BodyParser(&req); err != nil {
		return exception.InvalidRequestBodyError
	}

	authenticationValidation := validation.AuthenticationValidation{}
	if err := authenticationValidation.CredentialsValidation(req); err != nil {
		return err
	}

	user, accessToken, refreshToken, err := uc.authenticationService.RegisterUser(c.Context(), req)
	if err != nil {
		return err
	}

	auth.SetRefreshCookie(c, refreshToken)

	return c.Status(fiber.StatusCreated).JSON(response.CommonResponse{
		ResponseCode:    response.RESOURCE_CREATED,
		ResponseMessage: "User registered successfully",
		Data: dto.AuthTokensDTO{
			ID:          user.ID,
			Email:       user.Email,
			AccessToken: accessToken,
		},
	})
}

func (uc *Controller) Login(c *fiber.Ctx) error {
	var req dto.RegisterUserDTI

	if err := c.BodyParser(&req); err != nil {
		return exception.InvalidRequestBodyError
	}

	authenticationValidation := validation.AuthenticationValidation{}
	if err := authenticationValidation.CredentialsValidation(req); err != nil {
		return err
	}

	user, accessToken, refreshToken, err := uc.authenticationService.LoginUser(c.Context(), req)
	if err != nil {
		return err
	}

	auth.SetRefreshCookie(c, refreshToken)

	return c.Status(fiber.StatusOK).JSON(response.CommonResponse{
		ResponseCode:    response.RESOURCE_FETCHED,
		ResponseMessage: "Login successful",
		Data: dto.AuthTokensDTO{
			ID:          user.ID,
			Email:       user.Email,
			AccessToken: accessToken,
		},
	})
}

func (uc *Controller) Refresh(c *fiber.Ctx) error {
	rawToken := auth.GetRefreshCookie(c)
	if rawToken == "" {
		return exception.UnauthorizedError
	}

	accessToken, refreshToken, err := uc.authenticationService.RefreshTokens(c.Context(), rawToken)
	if err != nil {
		auth.ClearRefreshCookie(c)
		return err
	}

	auth.SetRefreshCookie(c, refreshToken)

	return c.Status(fiber.StatusOK).JSON(response.CommonResponse{
		ResponseCode:    response.RESOURCE_FETCHED,
		ResponseMessage: "Token refreshed",
		Data: dto.AuthTokensDTO{
			AccessToken: accessToken,
		},
	})
}

func (uc *Controller) Logout(c *fiber.Ctx) error {
	rawToken := auth.GetRefreshCookie(c)
	if rawToken != "" {
		_ = uc.authenticationService.Logout(c.Context(), rawToken)
	}

	auth.ClearRefreshCookie(c)

	return c.Status(fiber.StatusOK).JSON(response.CommonResponse{
		ResponseCode:    response.RESOURCE_UPDATED,
		ResponseMessage: "Logged out successfully",
	})
}

func (uc *Controller) ForgotPassword(c *fiber.Ctx) error {
	var req dto.ForgotPasswordDTI
	if err := c.BodyParser(&req); err != nil {
		return exception.InvalidRequestBodyError
	}

	authenticationValidation := validation.AuthenticationValidation{}
	if err := authenticationValidation.ForgotPasswordEmailValidation(req); err != nil {
		return err
	}

	if err := uc.authenticationService.SendPasswordResetEmail(c.Context(), req.Email); err != nil {
		return err
	}

	return c.Status(fiber.StatusOK).JSON(response.CommonResponse{
		ResponseCode:    response.RESOURCE_CREATED,
		ResponseMessage: "Password reset link has been sent.",
	})
}

func (uc *Controller) ResetPassword(c *fiber.Ctx) error {
	var req dto.ResetPasswordDTI
	if err := c.BodyParser(&req); err != nil {
		return exception.InvalidRequestBodyError
	}

	authenticationValidation := validation.AuthenticationValidation{}
	if err := authenticationValidation.ResetPasswordEmailValidation(req); err != nil {
		return err
	}

	if err := uc.authenticationService.ResetUserPassword(c.Context(), req); err != nil {
		return err
	}

	return c.Status(fiber.StatusOK).JSON(response.CommonResponse{
		ResponseCode:    response.RESOURCE_UPDATED,
		ResponseMessage: "Password has been reset successfully",
	})
}

func (uc *Controller) RequestEmailCode(c *fiber.Ctx) error {
	var req dto.RequestEmailCodeDTI
	if err := c.BodyParser(&req); err != nil {
		return exception.InvalidRequestBodyError
	}

	authenticationValidation := validation.AuthenticationValidation{}
	if err := authenticationValidation.RequestEmailCodeValidation(req); err != nil {
		return err
	}

	if err := uc.authenticationService.RequestLoginCode(c.Context(), req); err != nil {
		return err
	}

	return c.Status(fiber.StatusOK).JSON(response.CommonResponse{
		ResponseCode:    response.RESOURCE_FETCHED,
		ResponseMessage: "If that email is registered, a login code has been sent",
	})
}

func (uc *Controller) VerifyEmailCode(c *fiber.Ctx) error {
	var req dto.VerifyEmailCodeDTI
	if err := c.BodyParser(&req); err != nil {
		return exception.InvalidRequestBodyError
	}

	authenticationValidation := validation.AuthenticationValidation{}
	if err := authenticationValidation.VerifyEmailCodeValidation(req); err != nil {
		return err
	}

	user, accessToken, refreshToken, err := uc.authenticationService.VerifyLoginCode(c.Context(), req)
	if err != nil {
		return err
	}

	auth.SetRefreshCookie(c, refreshToken)

	return c.Status(fiber.StatusOK).JSON(response.CommonResponse{
		ResponseCode:    response.RESOURCE_FETCHED,
		ResponseMessage: "Login successful",
		Data: dto.AuthTokensDTO{
			ID:          user.ID,
			Email:       user.Email,
			AccessToken: accessToken,
		},
	})
}
